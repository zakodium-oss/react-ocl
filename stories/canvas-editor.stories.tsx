import {useCallback, useState} from 'react';

import CanvasEditor from '../src/components/CanvasEditor.js';
import type {Molecule} from "openchemlib/full";

export default {
    title: 'CanvasEditor',
    component: CanvasEditor,
    args: {
        mode: "molecule",
        fragment: false,
        width: 675,
        height: 450,
    },
    parameters: {
        docs: {
            description: {
                component: 'StructureEditor is an uncontrolled component.',
            },
        },
    },
};

const initialIDCode = 'gJX@@eKU@@ gFp@DiTt@@@!gGQHDHaImfh@##!B_vp@[G|S@AL !BmpJH@ox@?`BH?@ !Bb@K~@Hc}b@JH?P';

export function TheCanvasEditor()
{
    const [idCode, setIDCode] = useState(initialIDCode);
    const [previous, setPrevious] = useState<string | null>(null);
    const cb = useCallback(
        (molfile: string | null, molecule: Molecule | null, newIDCode: string | null) => {
            if (!newIDCode) {
                setIDCode("");
            } else {
                setIDCode(newIDCode);
            }
            setPrevious(idCode);
        },
        [setIDCode, setPrevious, idCode],
    );
    return (
        <div>
            <h2>Editor</h2>
            <CanvasEditor initialIDCode={idCode} onChange={cb} mode={"reaction"}/>
            <div style={{display: 'flex'}}>
                <div style={{flex: 1}}>
                    <h2 style={{textAlign: 'center'}}>Current</h2>
                    <pre>{idCode}</pre>
                </div>
                <div style={{flex: 1}}>
                    <h2 style={{textAlign: 'center'}}>Previous</h2>
                    <pre>{previous}</pre>
                </div>
            </div>
        </div>
    );
}
