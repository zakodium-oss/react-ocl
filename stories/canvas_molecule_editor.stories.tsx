import type { Meta, StoryObj } from '@storybook/react-vite';
import { useCallback, useState } from 'react';

import type { CanvasEditorOnChangeMolecule } from '../src/index.ts';
import { CanvasMoleculeEditor } from '../src/index.ts';

export default {
  title: 'CanvasMoleculeEditor',
  component: CanvasMoleculeEditor,
  args: {
    width: 675,
    height: 450,
    inputFormat: 'smiles',
    inputValue: 'CN1C2CCC1C(C(C2)OC(=O)C3=CC=CC=C3)C(=O)OC',
  },
} satisfies Meta<typeof CanvasMoleculeEditor>;

type Story = StoryObj<typeof CanvasMoleculeEditor>;

export const Basic: Story = {};

export const ReadOnly: Story = {
  args: {
    readOnly: true,
  },
};

export const Fragment: Story = {
  args: {
    fragment: true,
  },
};

export const Reactive: Story = {
  render(args) {
    const [data, setData] = useState({
      idcode: '',
      molfile: '',
      molfileV3: '',
      smiles: '',
    });
    const cb = useCallback((event: CanvasEditorOnChangeMolecule) => {
      setData({
        idcode: event.getIdcode(),
        molfile: event.getMolfile(),
        molfileV3: event.getMolfileV3(),
        smiles: event.getSmiles(),
      });
    }, []);
    return (
      <div>
        <h2>Reactive editor</h2>
        <p>
          The editor is not supposed to be used like this by always passing the
          state from the last event like a controlled component. This cannot
          work properly because of the way coordinates are managed internally.
          This story is here to show that it is possible to update the state by
          changing the input value, though.
        </p>
        <CanvasMoleculeEditor
          {...args}
          inputFormat="idcode"
          inputValue={data.idcode}
          onChange={cb}
        />
        <table>
          <tbody>
            <tr>
              <td valign="top">Idcode</td>
              <td>
                <code>{data.idcode}</code>
              </td>
            </tr>
            <tr>
              <td valign="top">Molfile V2000</td>
              <td>
                <code style={{ whiteSpace: 'pre' }}>{data.molfile}</code>
              </td>
            </tr>
            <tr>
              <td valign="top">Molfile V3000</td>
              <td>
                <code style={{ whiteSpace: 'pre' }}>{data.molfileV3}</code>
              </td>
            </tr>
            <tr>
              <td valign="top">SMILES</td>
              <td>
                <code>{data.smiles}</code>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    );
  },
};
