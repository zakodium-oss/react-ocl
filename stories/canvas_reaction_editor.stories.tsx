import type { Meta, StoryObj } from '@storybook/react';
import { useCallback, useState } from 'react';

import { CanvasReactionEditor } from '../src/components/canvas_editor.js';
import type { CanvasEditorOnChangeReaction } from '../src/components/canvas_editor_hook.js';

export default {
  title: 'CanvasReactionEditor',
  component: CanvasReactionEditor,
  args: {
    width: 800,
    height: 450,
    inputFormat: 'smiles',
    inputValue: 'CC(=O)O.OCC>>CC(=O)OCC',
  },
} satisfies Meta<typeof CanvasReactionEditor>;

type Story = StoryObj<typeof CanvasReactionEditor>;

export const Basic: Story = {};

export const ReadOnly: Story = {
  args: {
    readOnly: true,
  },
};

export const Reactive: Story = {
  render(args) {
    const [data, setData] = useState({
      idcode: '',
      rxn: '',
      rxnV3: '',
      smiles: '',
    });
    const cb = useCallback((event: CanvasEditorOnChangeReaction) => {
      setData({
        idcode: event.getIdcode(),
        rxn: event.getRxn(),
        rxnV3: event.getRxnV3(),
        smiles: event.getSmiles(),
      });
    }, []);
    return (
      <div>
        <h2>Reactive editor</h2>
        <p>
          The editor is not supposed to be used like this by always passing the
          state from the last event like a controlled component. This cannot
          work properly because of the way coordinates are managed internally.
          This story is here to show that it is possible to update the state by
          changing the input value, though.
        </p>
        <CanvasReactionEditor
          {...args}
          inputFormat="idcode"
          inputValue={data.idcode}
          onChange={cb}
        />
        <table>
          <tbody>
            <tr>
              <td valign="top">Idcode</td>
              <code>{data.idcode}</code>
            </tr>
            <tr>
              <td valign="top">RXN V2000</td>
              <code style={{ whiteSpace: 'pre' }}>{data.rxn}</code>
            </tr>
            <tr>
              <td valign="top">RXN V3000</td>
              <code style={{ whiteSpace: 'pre' }}>{data.rxnV3}</code>
            </tr>
            <tr>
              <td valign="top">SMILES</td>
              <td>
                <code>{data.smiles}</code>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    );
  },
};
