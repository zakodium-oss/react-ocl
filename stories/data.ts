import OCL from 'openchemlib/minimal';

export const molfileV2000 = `
  Marvin  07260611242D          

 24 26  0  0  0  0            999 V2000
   -0.6312    0.1434    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.1937    0.1434    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.5245    1.5927    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2250    0.5897    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2250    1.4147    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2187    1.9507    0.0000 C   0  0  1  0  0  0  0  0  0  0  0  0
   -0.9620    1.5927    0.0000 C   0  0  2  0  0  0  0  0  0  0  0  0
   -1.6765    2.8302    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -1.6765    2.0052    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.3910    1.5927    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -3.1055    2.0052    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2187    2.7757    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.4957    4.0132    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.4957    3.1882    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2102    2.7757    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.9247    1.5382    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.6391    1.9507    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.6391    2.7757    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.9247    3.1882    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2102    1.9507    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.1456    0.7885    0.0000 C   0  0  1  0  0  0  0  0  0  0  0  0
   -1.9425    0.5750    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.7081    0.7885    0.0000 C   0  0  1  0  0  0  0  0  0  0  0  0
    1.5050    0.5750    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  6  7  1  0  0  0  0
  3  6  1  0  0  0  0
  7 21  1  0  0  0  0
 21  1  1  0  0  0  0
  1  2  1  0  0  0  0
  2 23  1  0  0  0  0
 23  3  1  0  0  0  0
 21  4  1  1  0  0  0
 23  4  1  1  0  0  0
  4  5  1  0  0  0  0
  6 12  1  1  0  0  0
  7  9  1  1  0  0  0
  9  8  2  0  0  0  0
  9 10  1  0  0  0  0
 10 11  1  0  0  0  0
 12 14  1  0  0  0  0
 14 13  2  0  0  0  0
 14 15  1  0  0  0  0
 19 15  2  0  0  0  0
 15 20  1  0  0  0  0
 16 17  1  0  0  0  0
 20 16  2  0  0  0  0
 17 18  2  0  0  0  0
 18 19  1  0  0  0  0
 21 22  1  1  0  0  0
 23 24  1  1  0  0  0
M  END
`;

const mol = OCL.Molecule.fromMolfile(molfileV2000);

export const idcode = mol.getIDCodeAndCoordinates();

export const molfileV3000 = mol.toMolfileV3();
