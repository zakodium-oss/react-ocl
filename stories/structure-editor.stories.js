import { useCallback, useState } from 'react';

import StructureEditor from '../src/components/StructureEditor';

export default {
  title: 'StructureEditor',
  component: StructureEditor,
  args: {
    svgMenu: true,
    fragment: false,
    width: 675,
    height: 450,
  },
  parameters: {
    docs: {
      description: {
        component: 'StructureEditor is an uncontrolled component.',
      },
    },
  },
};

const initialIDCode = 'gFp@DiTvjh@ !B?g~w@k_}m?vw@`';

const initialMolfile = `
Actelion Java MolfileCreator 1.0

  6  5  0  0  0  0  0  0  0  0999 V2000
    3.4641   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5981   -0.0000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -1.5000   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.8660   -0.0000   -0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  2  1  1  0  0  0  0
  3  2  1  0  0  0  0
  4  3  2  0  0  0  0
  5  3  1  0  0  0  0
  6  5  1  0  0  0  0
M  END
`;

export function FromMolfile({ svgMenu, fragment, width, height }) {
  const [molfile, setMolfile] = useState(initialMolfile);
  const [previous, setPrevious] = useState(null);
  const cb = useCallback(
    (newMolfile) => {
      setMolfile(newMolfile);
      setPrevious(molfile);
    },
    [setMolfile, setPrevious, molfile],
  );
  return (
    <div>
      <h2>Editor</h2>
      <StructureEditor
        initialMolfile={molfile}
        {...{ svgMenu, width, height }}
        fragment={fragment}
        onChange={cb}
      />
      <div style={{ display: 'flex' }}>
        <div style={{ flex: 1 }}>
          <h2 style={{ textAlign: 'center' }}>Current</h2>
          <pre>{molfile}</pre>
        </div>
        <div style={{ flex: 1 }}>
          <h2 style={{ textAlign: 'center' }}>Previous</h2>
          <pre>{previous}</pre>
        </div>
      </div>
    </div>
  );
}

export function FromIDCode({ svgMenu, fragment, width, height }) {
  const [idCode, setIDCode] = useState(initialIDCode);
  const [previous, setPrevious] = useState(null);
  const cb = useCallback(
    (netMolfile, molecule, newIDCode) => {
      setIDCode(newIDCode);
      setPrevious(idCode);
    },
    [setIDCode, setPrevious, idCode],
  );
  return (
    <div>
      <h2>Editor</h2>
      <StructureEditor
        initialIDCode={idCode}
        {...{ svgMenu, width, height }}
        fragment={fragment}
        onChange={cb}
      />
      <div style={{ display: 'flex' }}>
        <div style={{ flex: 1 }}>
          <h2 style={{ textAlign: 'center' }}>Current</h2>
          <pre>{idCode}</pre>
        </div>
        <div style={{ flex: 1 }}>
          <h2 style={{ textAlign: 'center' }}>Previous</h2>
          <pre>{previous}</pre>
        </div>
      </div>
    </div>
  );
}
