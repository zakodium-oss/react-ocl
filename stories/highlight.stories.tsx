import { useState } from 'react';

import { MolfileSvgRenderer } from '../src/index.js';

import { molfileV2000 } from './data.js';

export default {
  title: 'Highlighting',
  args: {
    molfile: molfileV2000,
    atomHighlightColor: 'yellow',
    atomHighlightOpacity: 0.5,
    bondHighlightColor: 'red',
    bondHighlightOpacity: 0.5,
  },
  argTypes: {
    atomHighlightColor: {
      name: 'Atom highlight color',
      control: 'color',
    },
    atomHighlightOpacity: {
      name: 'Atom highlight opacity',
      control: 'number',
    },
    bondHighlightColor: {
      name: 'Bond highlight color',
      control: 'color',
    },
    bondHighlightOpacity: {
      name: 'Bond highlight opacity',
      control: 'number',
    },
  },
};

export function Fixed(args: any) {
  return <MolfileSvgRenderer {...args} />;
}
Fixed.storyName = 'Fixed highlight';
Fixed.args = {
  atomHighlight: [1, 5],
  bondHighlight: [6],
};
Fixed.argTypes = {
  atomHighlight: {
    name: 'Atom highlight ids',
    control: 'object',
  },
  bondHighlight: {
    name: 'Bond highlight ids',
    control: 'object',
  },
};

export function Hover(args: any) {
  const [currentAtom, setCurrentAtom] = useState(null);
  const [currentBond, setCurrentBond] = useState(null);
  return (
    <MolfileSvgRenderer
      {...args}
      atomHighlight={currentAtom ? [currentAtom] : null}
      onAtomEnter={setCurrentAtom}
      onAtomLeave={() => setCurrentAtom(null)}
      bondHighlight={currentBond ? [currentBond] : null}
      onBondEnter={setCurrentBond}
      onBondLeave={() => setCurrentBond(null)}
    />
  );
}
Hover.storyName = 'Highlight hovered element';
Hover.args = {};
