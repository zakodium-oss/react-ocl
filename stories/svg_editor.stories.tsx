import type { Meta, StoryObj } from '@storybook/react-vite';
import { useState } from 'react';

import { SvgEditor } from '../src/index.js';

import { molecule as defaultBaseMolecule } from './data.js';

const defaultMolecule = defaultBaseMolecule.getCompactCopy();
defaultMolecule.setAtomCustomLabel(0, 'Custom');

export default {
  title: 'SvgEditor',
  component: SvgEditor,
  // We cannot pass a Molecule as arg because it is not serializable.
  render: (args) => {
    const [molecule, setMolecule] = useState(defaultMolecule);

    return <SvgEditor {...args} molecule={molecule} onChange={setMolecule} />;
  },
} satisfies Meta<typeof SvgEditor>;

type Story = StoryObj<typeof SvgEditor>;

export const Control: Story = {};
