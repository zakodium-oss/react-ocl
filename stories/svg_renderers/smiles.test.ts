import { composeStories } from '@storybook/react-vite';
import { expect, test } from 'vitest';

import * as stories from './smiles.stories.tsx';

const { Smiles, DefaultError, CustomError } = composeStories(stories);

test('Molecule renders SMILES with custom id', async () => {
  await Smiles.run({ args: { ...Smiles.args, id: 'mol1' } });

  expect(document.body.firstChild).toMatchSnapshot();
});

test('Syntax error in SMILES - default renderer', async () => {
  await DefaultError.run();

  expect(document.body.firstChild).toMatchSnapshot();
});

test('Syntax error in SMILES - custom renderer', async () => {
  await CustomError.run();

  expect(document.body.firstChild).toMatchSnapshot();
});
