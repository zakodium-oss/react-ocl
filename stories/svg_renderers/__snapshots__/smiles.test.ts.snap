// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Molecule renders SMILES with custom id 1`] = `
<div>
  <svg
    height="100px"
    id="mol1"
    style="user-select: none;"
    version="1.1"
    viewBox="55 47 190 100"
    width="190px"
    xmlns="http://www.w3.org/2000/svg"
  >
    
  
    <style>
       text {font-family: sans-serif;} #mol1 { pointer-events:none; } #mol1 .event  { pointer-events:all; } line { stroke-linecap:round; } polygon { stroke-linejoin:round; } 
    </style>
    
  
    <text
      fill="rgb(255,13,13)"
      font-size="14"
      stroke="none"
      x="196.52"
      y="92.67"
    >
      O
    </text>
    
  
    <text
      fill="rgb(255,13,13)"
      font-size="14"
      stroke="none"
      x="134.16"
      y="80.67"
    >
      O
    </text>
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="118.82"
      x2="98.04"
      y1="88"
      y2="76"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="114.81"
      x2="98.46"
      y1="89.84"
      y2="80.4"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="77.25"
      x2="77.25"
      y1="88"
      y2="112"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="80.85"
      x2="80.85"
      y1="90.56"
      y2="109.44"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="98.04"
      x2="118.82"
      y1="124"
      y2="112"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="98.46"
      x2="114.81"
      y1="119.6"
      y2="110.16"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="222.75"
      x2="208.41"
      y1="76"
      y2="84.28"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="195.52"
      x2="181.18"
      y1="84.28"
      y2="76"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="181.18"
      x2="160.39"
      y1="76"
      y2="88"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="160.39"
      x2="146.05"
      y1="88"
      y2="79.72"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="133.16"
      x2="118.82"
      y1="79.72"
      y2="88"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="98.04"
      x2="77.25"
      y1="76"
      y2="88"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="77.25"
      x2="98.04"
      y1="112"
      y2="124"
    />
    
  
    <line
      stroke="rgb(0,0,0)"
      stroke-width="1.44"
      x1="118.82"
      x2="118.82"
      y1="88"
      y2="112"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:5"
      opacity="0"
      stroke-width="8"
      x1="118.82"
      x2="98.04"
      y1="88"
      y2="76"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:7"
      opacity="0"
      stroke-width="8"
      x1="77.25"
      x2="77.25"
      y1="88"
      y2="112"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:9"
      opacity="0"
      stroke-width="8"
      x1="98.04"
      x2="118.82"
      y1="124"
      y2="112"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:0"
      opacity="0"
      stroke-width="8"
      x1="222.75"
      x2="201.96"
      y1="76"
      y2="88"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:1"
      opacity="0"
      stroke-width="8"
      x1="201.96"
      x2="181.18"
      y1="88"
      y2="76"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:2"
      opacity="0"
      stroke-width="8"
      x1="181.18"
      x2="160.39"
      y1="76"
      y2="88"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:3"
      opacity="0"
      stroke-width="8"
      x1="160.39"
      x2="139.61"
      y1="88"
      y2="76"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:4"
      opacity="0"
      stroke-width="8"
      x1="139.61"
      x2="118.82"
      y1="76"
      y2="88"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:6"
      opacity="0"
      stroke-width="8"
      x1="98.04"
      x2="77.25"
      y1="76"
      y2="88"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:8"
      opacity="0"
      stroke-width="8"
      x1="77.25"
      x2="98.04"
      y1="112"
      y2="124"
    />
    
  
    <line
      class="event"
      id="mol1:Bond:10"
      opacity="0"
      stroke-width="8"
      x1="118.82"
      x2="118.82"
      y1="88"
      y2="112"
    />
    
  
    <circle
      class="event"
      cx="222.75"
      cy="76"
      id="mol1:Atom:0"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="201.96"
      cy="88"
      id="mol1:Atom:1"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="181.18"
      cy="76"
      id="mol1:Atom:2"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="160.39"
      cy="88"
      id="mol1:Atom:3"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="139.61"
      cy="76"
      id="mol1:Atom:4"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="118.82"
      cy="88"
      id="mol1:Atom:5"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="98.04"
      cy="76"
      id="mol1:Atom:6"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="77.25"
      cy="88"
      id="mol1:Atom:7"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="77.25"
      cy="112"
      id="mol1:Atom:8"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="98.04"
      cy="124"
      id="mol1:Atom:9"
      opacity="0"
      r="8"
    />
    
  
    <circle
      class="event"
      cx="118.82"
      cy="112"
      id="mol1:Atom:10"
      opacity="0"
      r="8"
    />
    

    <text
      fill="rgb(127,127,127)"
      font-family="Arial, Helvetica, sans-serif"
      font-size="14 "
      text-anchor="middle"
      x="150 "
      y="142.33333333333334 "
    >
      Label p1
    </text>
  </svg>
</div>
`;

exports[`Syntax error in SMILES - custom renderer 1`] = `
<div>
  <div
    style="width: 300px; height: 200px;"
  >
    <div
      style="color: red;"
    >
      <div>
        COVVVCC
      </div>
      <div>
        Class$S19: SmilesParser: unknown element label found. Position:2
      </div>
    </div>
  </div>
</div>
`;

exports[`Syntax error in SMILES - default renderer 1`] = `
<div>
  <div
    style="width: 300px; height: 200px;"
  >
    <svg
      height="200"
      version="1.1"
      viewBox="0 0 300 150"
      width="300"
      xmlns="http://www.w3.org/2000/svg"
    >
      <text
        fill="rgb(255,0,0)"
        font-size="15"
        x="0"
        y="2"
      >
        <tspan
          dy="10"
          x="0"
        >
          Invalid SMILES
        </tspan>
      </text>
    </svg>
  </div>
</div>
`;
